### PR TITLE
mountpoint-util: do not assume symlinks are not mountpoints

### PR DESCRIPTION
**Pull request by YHNdnzj**
They very much can be with the new mount API.

### CODE REVIEW

**Comment by YHNdnzj** [comment]
So I really think we should reconsider how we treat symlinks across the codebase. The outdated practice of assuming symlinks aren't mountpoints or always canonicalizing before mounting would no longer work well under the new mount API.

**Comment by poettering** [comment]
Hmm, I am pretty sure this is wrong. With the new mount api you can overmount symlinks, but you cannot overmount inodes with symlinks.

are you sure this actually works?

**Comment by YHNdnzj** [comment]
> Hmm, I am pretty sure this is wrong. With the new mount api you can overmount symlinks, but you cannot overmount inodes with symlinks.
> 
> are you sure this actually works?

Added a test case which illustrates how this would work.

File: src/test/test-mountpoint-util.c [hunk 1]
```diff
@@ -317,6 +318,37 @@ TEST(fd_is_mount_point) {
...
+                return;
+
+        /* Symlinks can be mount points with new mount API */
+        _cleanup_close_ int mfd = -EBADF, rfd = -EBADF;
+        _cleanup_free_ char *t = NULL;
+        struct stat st;
+
+        safe_close(fd);
+        fd = open(tmpdir, O_DIRECTORY|O_PATH|O_CLOEXEC);
+        assert_se(fd >= 0);
```
Comment (daandemeyer) [inline src/test/test-mountpoint-util.c]: `ASSERT_OK_ERRNO` should be used here.

File: src/test/test-mountpoint-util.c [hunk 2]
```diff
@@ -317,6 +318,37 @@ TEST(fd_is_mount_point) {
...
+
+        ASSERT_OK_ERRNO(symlinkat("/usr", fd, "symlink"));
+
+        mfd = open_tree(fd, "symlink", AT_SYMLINK_NOFOLLOW|OPEN_TREE_CLONE|OPEN_TREE_CLOEXEC);
+        if (mfd < 0 && ERRNO_IS_PRIVILEGE(errno))
+                return;
+        assert_se(mfd >= 0);
```
Comment (daandemeyer) [inline src/test/test-mountpoint-util.c]: Ditto

File: src/test/test-mountpoint-util.c [hunk 3]
```diff
@@ -317,6 +318,37 @@ TEST(fd_is_mount_point) {
...
+
+        rfd = openat(fd, "regular", O_CLOEXEC|O_CREAT|O_EXCL, 0644);
+        assert_se(rfd >= 0);
```
Comment (daandemeyer) [inline src/test/test-mountpoint-util.c]: Ditto

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

**Review by bluca — APPROVED** [review]

**Pull request merged**